### DISCUSSION TITLE
PHP and ImageMagick (status feedback: 'complete', 'done', 'finished', etc)

### DISCUSSION THREAD

**Discussion by anthonypaulbalducci** [ImageMagick/ImageMagick#5665] [Help]
Hmm, so I did a general Google search on this using various search terms, but was surprisingly not able to find anything relevant, nor in the docs. 

I'm calling up ImageMagick from PHP on a webserver, but running into a bit of an issue insofar as my next page reference I think sometimes is passing on to the next page (which requires the edited and outputted image from ImageMagick), but sometimes before IM has actually finished processing/writing the new file.

I mean I suppose in PHP I could do some sort of loop for 'file_exists'-- Though of course just because a 'file exists' doesn't yet mean it is 'fully written'.

Thus, I just wondered does IM provide any 'feedback' that is accessible to PHP to say, of any given operation: 'Ok, I'm done !' (Or even, alternatively, 'operation failed') ?

Best,
-A

**Comment by fmw42** [comment]
Imagemagick outputs an exit code, 0 or 1 when finished. You might be able to do an if test on the exit code.  Sorry, I only do bash scripting (PHP exec) not Imagick, if that is what you are using.

**Reply by anthonypaulbalducci** [reply]
Oh, Hey Fred, thank you for providing me with the right 'grammar' to use. I would have never suspected 'exit code' as the term to use, but it is good to know it outputs a 'success/fail' result. I have found some examples of 'catching this' if you are using 'exec'.. But I'm not, I am just using it 'in-line'. In any case, once I determine a possible solution, I will post it as an answer here.

Best,
-A

**Reply by fmw42** [reply]
What do you mean by using it "in-line"?  Are you saying in the command line?  Or are you using Imagick?  What tool or software?

**Comment by snibgo** [comment]
@anthonypaulbalducci: Please say what product you are using. Is it IMagick, or ImageMagick? This is an ImageMagick forum.

Please show your code. Without seeing that, advice is difficult. But if it is IMagick code, you might find an IMagick forum is more useful.

**Reply by anthonypaulbalducci** [reply]
Dear Snibgo,

I am sorry to sound ignorant, but until very recently I kind of 'assumed' that ImageMagick and Imagick were the 'same thing'-- Only just a 'server-side' implementation in PHP added in as an extension.

As to the 'exact code' in this case (and yeah, yeah, I need to clean it up):

```
`<?php
header("Access-Control-Allow-Origin: *"); // allow all domains, FOR TESTING
// start a session
session_start();

// A list of permitted file extensions
$allowed = array('png', 'jpg', 'gif');

if(isset($_FILES['upl']) && $_FILES['upl']['error'] == 0){

	$extension = pathinfo($_FILES['upl']['name'], PATHINFO_EXTENSION);

	if(!in_array(strtolower($extension), $allowed)){
		echo '{"status":"error"}';
		exit;
	}

	if(move_uploaded_file($_FILES['upl']['tmp_name'], 'uploads/'.$_FILES['upl']['name'])){
       // header("Location: https://www.allpointsbulletin.info", TRUE, 301);
		echo '{"status":"success"}';
        $_SESSION['front_image'] = $_FILES['upl']['name'];
        $outFile = $_FILES['upl']['name'];
        $image = new Imagick('uploads/'.$_FILES['upl']['name']);
        $image->thumbnailImage(450, 450, true);
        $image->writeImage($outFile);
        rename($outFile, 'uploads/'.'TN_'.$outFile);

		exit;
	}
}

echo '{"status":"error"}';
exit;
`

```
But you can try this out at www.allpointsbulletin.info/test/

Upload an image, and *sometimes* unfortunately it will give you just an empty image reference on the next page. *But* if you simply hit 'refresh' on that page, you will see the image appears.

As of right now, I'm doing the 'page redirect' after uploading via JS, but I really think it should be instead done via PHP-- *If* I can figure out how Imagick has 'completed it's task'.

But, like I said, Fred has at least provided me with a 'good start'.

And just to answer/respond to his question about 'in-line'... The commands are directly in the PHP, or I am not relying on a second (i.e. BASH or shell script) to run them.

Anyways, sorry, still a little bit of a 'noobie'.....

Best,
-A.
```
```

**Reply by fmw42** [reply]
> And just to answer/respond to his question about 'in-line'... The commands are directly in the PHP, or I am not relying on a second (i.e. BASH or shell script) to run them.


Sorry, but I do not see any PHP exec commands that run Imagemagick directly outside of PHP Imagick.   Am I misunderstanding something here?

**Comment by snibgo** [comment]
Does the "`rename`" succeed? You don't check the return value. If it fails, you might wait a second and try again.

**Comment by anthonypaulbalducci** [comment]
Dear Fred,

No, I do not feel you are misunderstanding anything here. With a bit more
research on my part, I guess I thought Imagemagick == Imagick, but it is
kind of more an API for the library that is 'built-in' to modern PHP.

Yes, I follow you, if I was running an 'exec', on a shell, I could
immediately grab that 'exit code'.

I must admit, the most common problem I've had with my developing app is
that going to the 'resize' page, it will present, when it fails, a 'null
image', or 'TN_' (for 'thumbnail') but without the proper file name of the
image attached to it.

It seems this is kind of a 'race condition', and I am not sure yet if,
somehow PHP is not updating my $_SESSION variables fast enough, *or* I am
waiting on Imagick to finish processing.

I do quite understand, as snigbo said, this might tinge on a better
question for another forum.

Yet, I am quite pleased you provided the 'exit code' grammar.

Your inputs are always very helpful. Give me a little time to figure this
out and I will get back to you.

Best,
-A
